common/configuration/asm.go

```diff
@@ -0,0 +1,211 @@
+package configuration
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+
+	"github.com/TBD54566975/ftl/internal/slices"
+
+	. "github.com/alecthomas/types/optional" //nolint:stylecheck
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/credentials"
+	"github.com/aws/aws-sdk-go-v2/service/secretsmanager"
+	"github.com/aws/aws-sdk-go-v2/service/secretsmanager/types"
+	"github.com/aws/smithy-go"
+)
+
+// ASM implements Resolver and Provider for AWS Secrets Manager (ASM).
+//
+// The resolver does a direct/proxy map from a Ref to a URL, module.name <-> asm://module.name and does not access ASM at all.
+type ASM[R Secrets] struct {
+	client secretsmanager.Client
+}
+
+var _ Resolver[Secrets] = &ASM[Secrets]{}
+var _ Provider[Secrets] = &ASM[Secrets]{}
+var _ MutableProvider[Secrets] = &ASM[Secrets]{}
+
+func asmURLForRef(ref Ref) *url.URL {
+	return &url.URL{
+		Scheme: "asm",
+		Host:   ref.String(),
+	}
+}
+
+// NewASMWithDefaultCredentials creates a new ASM resolver/provider that uses the default AWS SDK credentials chain.
+func NewASMWithDefaultCredentials(ctx context.Context) (*ASM[Secrets], error) {
+	return NewASM(ctx, None[string](), None[string](), None[string](), None[string]())
+}
+
+// NewASM creates a new ASM resolver/provider with optional access key, secret access key, region, and endpoint.
+func NewASM(ctx context.Context, accessKeyID, secretAccessKey, region, endpoint Option[string]) (*ASM[Secrets], error) {
```

Comment:
That is a lot of optionals, and it's not clear what needs to be provided with what. Maybe this should just take an AWS credentials type (can't remember what it is exactly) directly?

Reply:
As discussed, pass in the client, and now we don't need the constructor again!

Reply:
🎢